service/history/ndc_task_util.go

```diff
@@ -243,9 +243,24 @@ func validateTaskByClock(
 	return nil
 }
 
-func validateTaskGeneration(mutableState workflow.MutableState, taskID int64) error {
+func validateTaskGeneration(
+	ctx context.Context,
+	shardContext shard.Context,
+	workflowContext workflow.Context,
+	mutableState workflow.MutableState,
+	taskID int64,
+) error {
 	tgClock := mutableState.GetExecutionInfo().TaskGenerationShardClockTimestamp
 	if tgClock != 0 && taskID != 0 && taskID < tgClock {
+
+		currentClock := shardContext.CurrentVectorClock().Clock
+		if tgClock >= currentClock {
```

Comment:
NOTE: this can only catch some cases, not all of them, as current clock is keep moving forward. Still better than not having it though I think.